### ISSUE TITLE
'finally' clause not accounted for in control flow analysis

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#11307]
Found in https://github.com/jquery/esprima/pull/1572/files/3595bf6dd293fe8275e950af3b93ef94b1889ea6 (@ariya)

Start out with the following with `noImplicitReturns`:

```ts
function f() {
    let x = 100;
    try {
        throw "WAT"
    }
    finally {
        return x;
    }
}
```

Now add a `catch` clause

```ts
function f() {
    let x = 100;
    try {
        throw "WAT"
    }
    catch (e) {
        
    }
    finally {
        return x;
    }
}
```

Expected: No errors on either.
Actual: Sample with the catch clause claims that all paths are exhaustive.
